File: app/src/main/java/com/killain/organizer/packages/fragments/TasksFragment.java
package com.killain.organizer.packages.fragments;

import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.design.widget.BottomNavigationView;
import android.support.v4.app.Fragment;
import android.support.v7.widget.RecyclerView;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.FrameLayout;
import android.widget.RelativeLayout;
import android.widget.ScrollView;
import android.widget.TextView;

import com.github.clans.fab.FloatingActionButton;
import com.killain.organizer.packages.enums.AdapterRefreshType;
import com.killain.organizer.packages.enums.DialogType;
import com.killain.organizer.packages.interactors.RecyclerViewInteractor;
import com.killain.organizer.packages.interactors.UIInteractor;
import com.killain.organizer.packages.models.Task;
import com.killain.organizer.packages.recyclerview_adapters.RVCardAdapter;
import com.killain.organizer.R;
import com.killain.organizer.packages.interfaces.FragmentUIHandler;
import org.threeten.bp.LocalDate;

import java.util.Calendar;

public class TasksFragment extends Fragment implements FragmentUIHandler {

    private int oldScrollYPosition = 0;
    public TextView noTaskTxt;
    public RelativeLayout relative_layout;
    public FrameLayout tasks_frame_layout, dialog_frame_layout;
    public FloatingActionButton fab_simple_task;
    private BottomNavigationView bottomNavigationView;
    public ScrollView scrollView;
    public RecyclerView recyclerView;
    public RVCardAdapter adapter;
    public TasksFragment tasksFragmentInstance;
    private UIInteractor uiInteractor;
    private LocalDate localDate;

    public TasksFragment() {
    }

    public static TasksFragment newInstance() {
        return new TasksFragment();
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        tasksFragmentInstance = TasksFragment.this;
        localDate = LocalDate.now();
    }

    @Override
    public void onStart() {
        if (adapter != null) {
            refreshAdapterOnAdd(0, AdapterRefreshType.RELOAD_FROM_DB);
        }
        super.onStart();
    }

    @Override
    public void onResume() {
        if (adapter != null) {
            refreshAdapterOnAdd(0, AdapterRefreshType.RELOAD_FROM_DB);
        }
        super.onResume();
    }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {

        View RootView = inflater.inflate(R.layout.fragment_tasks, container, false);

        noTaskTxt = RootView.findViewById(R.id.no_task_txt);
        fab_simple_task = getActivity().findViewById(R.id.fab_simple_task);
        bottomNavigationView = getActivity().findViewById(R.id.navigation);
        bottomNavigationView.getMenu().getItem(1).setChecked(true);
        tasks_frame_layout = getActivity().findViewById(R.id.content_frame_layout);
        dialog_frame_layout = getActivity().findViewById(R.id.dialog_frame_layout);
        scrollView = RootView.findViewById(R.id.tasks_frg_scroll_view);
        relative_layout = RootView.findViewById(R.id.parent_layout_tasks_fragment);
        recyclerView = RootView.findViewById(R.id.recycler_fragment_tasks);

        uiInteractor = new UIInteractor(relative_layout, fab_simple_task, bottomNavigationView);
        recyclerView.setNestedScrollingEnabled(false);
        RecyclerViewInteractor rvInteractor = new RecyclerViewInteractor(recyclerView);
        adapter = new RVCardAdapter (getContext(), rvInteractor.getListener(), this);
        rvInteractor.setAdapter(adapter);
        adapter.loadItemsByState();
        adapter.setParentFragment(tasksFragmentInstance);
        rvInteractor.bind();

        fab_simple_task.setOnClickListener(v -> {
            AddTaskDialogFragment dialog = new AddTaskDialogFragment();
            dialog.setDialogType(DialogType.ADD_NEW_TASK);
            dialog.setListener(this);
            dialog.setDate(localDate);
            if (getFragmentManager() != null) {
                dialog.show(getFragmentManager(), "dialog");
            }
            tasksFragmentInstance.onPause();
        });

        scrollView.getViewTreeObserver().addOnScrollChangedListener(() -> {
            if (scrollView.getScrollY() > oldScrollYPosition) {
                fab_simple_task.hide(true);
            } else if (scrollView.getScrollY() < oldScrollYPosition || scrollView.getScrollY() <= 0) {
                fab_simple_task.show(true);
            }
            oldScrollYPosition = scrollView.getScrollY();
        });

        return RootView;
    }

    @Override
    public View getBackground() {
        return relative_layout;
    }

    @Override
    public void refreshAdapterOnAdd(int position, AdapterRefreshType refreshType) {
        if (refreshType == AdapterRefreshType.RELOAD_FROM_DB) {
            adapter.loadItemsByState();
            adapter.notifyDataSetChanged();
        } else {
            adapter.notifyItemInserted(position);
        }
    }

    public void callDialogFragment(DialogType dialogType,
                                   FragmentUIHandler fragmentUIHandler,
                                   long date,
                                   Task task) {

        AddTaskDialogFragment dialog = new AddTaskDialogFragment();
        dialog.setDialogType(dialogType);
        dialog.setTaskAndDate(task, date);
        dialog.setListener(fragmentUIHandler);
        if (getFragmentManager() != null) {
            dialog.show(getFragmentManager(), "dialog");
        }
        tasksFragmentInstance.onPause();
    }

    @Override
    public void refreshAdapterOnDelete(int position) {
        adapter.notifyItemRemoved(position);
    }

    @Override
    public void onDestroy() {
        adapter.onDestroy();
        super.onDestroy();
    }

    @Override
    public void onPause() {
        adapter.onDestroy();
        super.onPause();
    }
}

File: app/src/main/java/com/killain/organizer/packages/fragments/TasksFragment.kt
package com.killain.organizer.packages.fragments

import android.os.Bundle
import android.support.design.widget.BottomNavigationView
import android.support.v4.app.Fragment
import android.support.v7.widget.RecyclerView

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.FrameLayout
import android.widget.RelativeLayout
import android.widget.ScrollView
import android.widget.TextView

import com.github.clans.fab.FloatingActionButton
import com.killain.organizer.packages.enums.AdapterRefreshType
import com.killain.organizer.packages.enums.DialogType
import com.killain.organizer.packages.interactors.RecyclerViewInteractor
import com.killain.organizer.packages.interactors.UIInteractor
import com.killain.organizer.packages.models.Task
import com.killain.organizer.packages.recyclerview_adapters.RVCardAdapter
import com.killain.organizer.R
import com.killain.organizer.packages.interfaces.FragmentUIHandler
import org.threeten.bp.LocalDate

class TasksFragment : Fragment(), FragmentUIHandler, View.OnClickListener {

    private var oldScrollYPosition = 0
    private lateinit var noTaskTxt: TextView
    private lateinit var relativeLayout: RelativeLayout
    private lateinit var tasksFrameLayout: FrameLayout
    private lateinit var dialogFrameLayout: FrameLayout
    private lateinit var fabSimpleTask: FloatingActionButton
    private var bottomNavigationView: BottomNavigationView? = null
    private lateinit var scrollView: ScrollView
    private lateinit var recyclerView: RecyclerView
    var adapter: RVCardAdapter? = null
    private lateinit var tasksFragmentInstance: TasksFragment
    private var uiInteractor: UIInteractor? = null
    private var localDate: LocalDate? = null

    override val background: View?
        get() = relativeLayout

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        tasksFragmentInstance = this
        localDate = LocalDate.now()
    }

    override fun onStart() {
        if (adapter != null) {
            refreshAdapterOnAdd(0, AdapterRefreshType.RELOAD_FROM_DB)
        }
        super.onStart()
    }

    override fun onResume() {
        if (adapter != null) {
            refreshAdapterOnAdd(0, AdapterRefreshType.RELOAD_FROM_DB)
        }
        super.onResume()
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {

        val rootView = inflater.inflate(R.layout.fragment_tasks, container, false)

        noTaskTxt = rootView.findViewById(R.id.no_task_txt)
        fabSimpleTask = activity!!.findViewById(R.id.fab_simple_task)
        bottomNavigationView = activity!!.findViewById(R.id.navigation)
        bottomNavigationView!!.menu.getItem(1).isChecked = true
        tasksFrameLayout = activity!!.findViewById(R.id.content_frame_layout)
        dialogFrameLayout = activity!!.findViewById(R.id.dialog_frame_layout)
        scrollView = rootView.findViewById(R.id.tasks_frg_scroll_view)
        relativeLayout = rootView.findViewById(R.id.parent_layout_tasks_fragment)
        recyclerView = rootView.findViewById(R.id.recycler_fragment_tasks)

        uiInteractor = UIInteractor(relativeLayout, fabSimpleTask, bottomNavigationView)
        recyclerView.isNestedScrollingEnabled = false
        val rvInteractor = RecyclerViewInteractor(recyclerView)
        adapter = RVCardAdapter(context!!, rvInteractor.listener, this)
        rvInteractor.adapter = adapter
        adapter?.loadItemsByState()
        adapter?.fragment = tasksFragmentInstance
        rvInteractor.bind()

        fabSimpleTask.setOnClickListener(this)

        scrollView.viewTreeObserver.addOnScrollChangedListener {
            if (scrollView.scrollY > oldScrollYPosition) {
                fabSimpleTask.hide(true)
            } else if (scrollView.scrollY < oldScrollYPosition || scrollView.scrollY <= 0) {
                fabSimpleTask.show(true)
            }
            oldScrollYPosition = scrollView.scrollY
        }

        return rootView
    }

    override fun refreshAdapterOnAdd(position: Int, refreshType: AdapterRefreshType) {
        if (adapter != null) {
            if (refreshType === AdapterRefreshType.RELOAD_FROM_DB) {
                adapter?.loadItemsByState()
                adapter?.notifyDataSetChanged()
            } else {
                adapter?.notifyItemInserted(position)
            }
        } else {
            return
        }
    }

    override fun refreshAdapterOnDelete(position: Int) {
        if (adapter != null) {
            adapter?.notifyItemRemoved(position)
        }
    }

    fun callDialogFragment(dialogType: DialogType,
                           fragmentUIHandler: FragmentUIHandler?,
                           date: Any?,
                           task: Task?) {

        val dialog = AddTaskDialogFragment()
        dialog.setDialogType(dialogType)
        dialog.setParams(task, date)
        dialog.setListener(fragmentUIHandler)
        if (fragmentManager != null) {
            dialog.show(fragmentManager, "dialog")
        }
        tasksFragmentInstance.onPause()
    }

    override fun onClick(v: View) {
        //        AddTaskDialogFragment dialog = new AddTaskDialogFragment();
        //        dialog.setDialogType(DialogType.ADD_NEW_TASK);
        //        dialog.setListener(this);
        //        dialog.setDate(localDate);
        //        if (getFragmentManager() != null) {
        //            dialog.show(getFragmentManager(), "dialog");
        //        }
        //        tasksFragmentInstance.onPause();
        callDialogFragment(DialogType.ADD_NEW_TASK, this, localDate, null)
    }

    override fun onDestroy() {
        adapter!!.onDestroy()
        super.onDestroy()
    }

    override fun onPause() {
        adapter!!.onDestroy()
        super.onPause()
    }

    companion object {
        fun newInstance(): TasksFragment {
            return TasksFragment()
        }
    }
}